plugins/vbe/fu-plugin-vbe.c

```diff
@@ -0,0 +1,279 @@
+fu_plugin_vbe_init(FuPlugin *plugin)
+{
+	FuPluginData *priv;
+	(void)fu_plugin_alloc_data(plugin, sizeof(FuPluginData));
+	priv = fu_plugin_get_data(plugin);
+	priv->vbe_dir = NULL;
+	priv->methods = NULL;
+}
+
+static void
+fu_plugin_vbe_destroy(FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data(plugin);
+	if (priv->vbe_dir)
+		g_free(priv->vbe_dir);
+}
+
+static gboolean vbe_locate_device(gchar *fdt, int node,
+				  struct FuVbeMethod **methp)
+{
+	struct FuVbeMethod *meth = NULL;
+	const struct VbeDriver *driver;
+	const gchar *method_name;
+	const char *compat, *p;
+
+	int len;
+
+	compat = fdt_getprop(fdt, node, "compatible", &len);
+	if (!compat) {
+		g_error("Missing update mechanism (%s)", fdt_strerror(len));
+		return FALSE;
+	}
+	p = strchr(compat, ',');
+	if (!p) {
+		g_error("Invalid update mechanism (%s)", compat);
+		return FALSE;
+	}
+	method_name = p + 1;
+
+	/* find this update mechanism */
+	for (driver = driver_list; driver->name; driver++) {
+		if (!strcmp(method_name, driver->name)) {
+			meth = g_malloc(sizeof(struct FuVbeMethod));
+			meth->vbe_method = p + 1;
+			meth->node = node;
+			meth->driver = driver;
+			g_info("Update mechanism: %s", meth->vbe_method);
+			*methp = meth;
+			return TRUE;
+		}
+	}
+	g_error("No driver for VBE method '%s'", method_name);
+
+	return FALSE;
+}
+
+static gboolean process_system(FuPluginData *priv, gchar *fdt, gsize fdt_len,
+			       GError **error)
+{
+	int ret, parent, node;
+	int found;
+
+	ret = fdt_check_header(fdt);
+	if (ret) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			    "System DT is corrupt (%s)",
+			    fdt_strerror(ret));
+		return FALSE;
+	}
+	if (fdt_totalsize(fdt) != fdt_len) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			    "System DT size mismatch (header=%x, file=%zx)",
+			    fdt_totalsize(fdt), fdt_len);
+		return FALSE;
+	}
+	parent = fdt_path_offset(fdt, NODE_PATH);
+	if (parent < 0) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			    "Missing node '%s' (%s)", NODE_PATH,
+			    fdt_strerror(ret));
+		return FALSE;
+	}
+
+	/* Create a device for each subnode */
+	found = 0;
+	for (node = fdt_first_subnode(fdt, parent); node > 0;
+	     node = fdt_next_subnode(fdt, node)) {
+		struct FuVbeMethod *meth;
+
+		if (vbe_locate_device(fdt, node, &meth)) {
+			found++;
+		} else {
+			g_warning("Cannot locate device for node '%s'",
+				  fdt_get_name(fdt, node, NULL));
+		}
+		priv->methods = g_list_append(priv->methods, meth);
+	}
+
+	if (found) {
+		g_info("VBE update methods: %d", found);
+	} else {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED,
+			    "No valid VBE update mechanism found");
+	}
+
+
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_vbe_startup(FuPlugin *plugin, FuProgress *progress, GError **error)
+{
+	FuPluginData *priv = fu_plugin_get_data(plugin);
+	g_autofree gchar *vbe_dir = NULL;
+	g_autofree gchar *state_dir = NULL;
+	gchar *buf, *bfname;
```

Comment:
You should be using `g_autofree` for `bfname` and initialize this to NULL.

Reply:
done